### PR TITLE
Copter: Excluded target velocity from slowdown calculations in FOLLOW mode

### PR DESCRIPTION
**Pull request by Georacer**
# Summary

This PR fixes a bug where Copter would not reach its target in FOLLOW mode.

# The issue

Currently, when the Copter would be in FOLLOW mode, following a **moving** target, it would never actually reach its desired position; instead there would be a steady-state position error.

The reason is that there are slowdown calculations in place in `ArduCopter/mode_follow.cpp`, in order to overwrite the position controller velocity setpoint and slow down the UA as it approaches its target, and prevent an overshoot.
However, these slowdown calculations would be applied over the absolute velocity, not the velocity relative to the target.
Effectively, the absolute velocity would be required to be zero above the moving target, which is inconsistent with a moving target.

This issue likely blocks the implementation of an effective Copter landing on a moving base.

# The changes

## Before

`ArduCopter/mode_follow.cpp:L65` and beyond carries out the calculations for the desired velocity.
In order, the steps taken would be:

1. Get the position offset from the target. (`L67`)
2. Use `FOLL_POS_P` gain to calculate approach velocity **and add target velocity**. (`L69-73`)
3. Apply the maximum velocity limits. (`L75-85`)
4. Carry out the slow-down calculations to limit lateral velocity based on distance from target. (`L94-112`).

## After

With this PR, the order of the steps is slightly changed.

1. Get the position offset from the target. (`L67`)
2. Use `FOLL_POS_P` gain to calculate approach velocity **without adding** target velocity. (`L69-73`). This yields the desired relative approach velocity.
3. Carry out the slow-down calculations to limit lateral velocity based on distance from target. (`L75-93`)
4. Add the target velocity to obtain the absolute desired velocity. (`L95-98`)
5. Apply the maximum velocity limits. (`L100-110`)

# Testing

A following scenario similar to that of https://github.com/ArduPilot/ardupilot/pull/24720 was simulated.
A truck would drive around at ~6m/s and the Copter was tasked to follow it.

Before, the Copter could get no closer than ~15m.
Screenshot from 2024-02-04 23-01-19

After, the Copter could effectively nullify the tracking error:
Screenshot from 2024-02-04 23-02-11

The simulated flight logs are available upon request. Let me know where I should upload them.

:information_source:  A script was running during the tests, but I don't think it alters the presented results.

### CODE REVIEW

**Comment by IamPete1** [comment]
This seems like a slightly different version of https://github.com/ArduPilot/ardupilot/pull/26138? This does have documented testing tho, so points for that ;)

**Comment by Georacer** [comment]
@IamPete1 well, two people tried to solve the same problem at the same day, after years of this code being live. What are the odds!

The only comment-worthy difference with @tridge 's code is that in his PR he applies the maximum speed limit (e.g. `WPNAV_SPEED` or `RTL_SPEED`) **before** adding the target's velocity. So there's a good chance the UA will actually overspeed and fail to limit itself to the specified value.

Other than that, tridge's PR has some nicer vector math.

**Comment by tridge** [comment]
> The only comment-worthy difference with @tridge 's code is that in his PR he applies the maximum speed limit (e.g. `WPNAV_SPEED` or `RTL_SPEED`) **before** adding the target's velocity. So there's a good chance the UA will actually overspeed and fail to limit itself to the specified value.

I fixed that in my PR this morning :-)

**Comment by IamPete1** [comment]
@Georacer Can you review the latest version of https://github.com/ArduPilot/ardupilot/pull/26138, We think the maths should now be the same?

**Comment by Georacer** [comment]
> @Georacer Can you review the latest version of #26138, We think the maths should now be the same?

@IamPete1 I have done so. I have requested some changes though, it's not 100% equivalent yet.

**Review by peterbarker — DISMISSED** [review]
Commit message needs fixing

**Comment by Georacer** [comment]
> Commit message needs fixing

@peterbarker I re-wrote it. Is this what you expected? I went by this guideline: https://ardupilot.org/dev/docs/submitting-patches-back-to-master.html

**Comment by peterbarker** [comment]
> > Commit message needs fixing
> 
> @peterbarker I re-wrote it. Is this what you expected? I went by this guideline: https://ardupilot.org/dev/docs/submitting-patches-back-to-master.html

Yep, that's good, thanks.

File: ArduCopter/mode_follow.cpp [hunk 1]
```diff
@@ -104,13 +83,27 @@ void ModeFollow::run()
...
         // limit vertical desired_velocity_neu_cms to slow as we approach target (we use 1/2 of maximum deceleration for gentle slow down)
         const float des_vel_z_max = copter.avoid.get_max_speed(pos_control->get_pos_z_p().kP().get(), pos_control->get_max_accel_z_cmss() * 0.5f, fabsf(dist_vec_offs_neu.z), copter.G_Dt);
         desired_velocity_neu_cms.z = constrain_float(desired_velocity_neu_cms.z, -des_vel_z_max, des_vel_z_max);
 
+        // Add the target velocity baseline.
+        desired_velocity_neu_cms.xy() += vel_of_target.xy() * 100.0f;
+        desired_velocity_neu_cms.z += -vel_of_target.z * 100.0f;
+
+        // scale desired velocity to stay within horizontal speed limit
+        float desired_speed_xy = safe_sqrt(sq(desired_velocity_neu_cms.x) + sq(desired_velocity_neu_cms.y));
```
Comment (tridge) [inline ArduCopter/mode_follow.cpp:97]: I think we can replace this block with:
```
  desired_velocity_neu_cms.xy().limit_length(pos_control->get_max_speed_xy_cms());
```

**Review by tridge — CHANGES_REQUESTED** [review]

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
I tested in SITL, looks good!

**Comment by rmackay9** [comment]
Merged, thanks!

**Pull request merged**